kubernetes/cluster-dump-restore.md

```diff
@@ -0,0 +1,91 @@
+
+_The deployment state of the Kubernetes cluster is stored in etcd. If you're concerned about backing up this information, you should look into backing up the etcd data directory for each etcd instance in your cluster. This can be done via an etcd-based backup strategy [example](https://github.com/coreos/etcd/tree/master/contrib/systemd/etcd2-backup-coreos), or via snapshotting the underlying block device that backs the etcd data directory. Backing up Kubernetes clusters is not the purpose of this document._
+
+This document's primary purpose is to instruct how to migrate the deployment state from one Kubernetes cluster to another. The clusters may have different versions, pod/service network cidrs, number of nodes, etc.
+
+For the remainer of this document, the cluster that is being dumped will be referred to as the *source cluster*. The cluster that is being restored to will be called the *target cluster*. The goal is to migrate state from the *source cluster* to the *target cluster*.
+
+This migration process boils down to selectively omitting state that is not portable between clusters, as well as making some assumptions about the clusters we're dealing with. Those assumptions include:
+* **kube-dns add-on** This allows abstracting away the services' clusterIPs, which is inherentely non-portable across clusters with different serviceCIDRS.
+
+* **api object compatiblilty** Any api object type (eg: replicaset) that is dumped from the *source cluster* must be supported by the *target cluster*.
+
+	As a general rule, if the *target cluster's* Kubernetes version is >= *source cluster's* Kubernets version, this will not be a problem.
+
+* **blank target slate** The *target cluster* to has no deployment state beyond a "stock" cluster. In the CoreOS case, the stock cluster state is everything in the `kube-system` namespace along with the service token in the default namespace.
+
+	If the *target cluster* has other pre-existing state, more precautions must be taken to ensure there are no undesired interactions between pre-existing resources on the *target cluster* and the dump coming from *source cluster*.
+
+	Name conflicts are easy to defect, but more subtle issues like services selecting unintended pre-existing pods is another more subtle situation that can arise. We will assume **blank target slate** for the remainder of this document.
+
+As of now, this is not entirely supported upstream. This `get --export` flag is supposed to do something like this, but leaves plenty of state that we do not consider to be portable in this case. (TODO: link to issue).
+
+We'll explain step-by-step how the CoreOS infrastructure team migrates deployments between Kubernets clusters today.
+
+### Dump from the source cluster
+
+Create a directory to hold your dump files:
+
+```shell
+mkdir ./cluster-dump
+```
+
+
+Get a list all namespaces that are not `kube-system` or `default`. This represents the list of namespaces that we want to migrate:
```

Comment:
We should mention that we're writing these out to a file, for folks that might not be super familiar with more advanced bash usage. Maybe:

> **First,** get a list all namespaces that are not `kube-system` or `default` **and record them into a file on disk.**